cpu/sam0_common/periph/freqm.c

```diff
@@ -0,0 +1,224 @@
+{
+    GCLK->GENCTRL[id].reg = GCLK_GENCTRL_SRC(src) | GCLK_GENCTRL_GENEN | flags | GCLK_GENCTRL_IDC;
+    while (GCLK->SYNCBUSY.reg & GCLK_SYNCBUSY_GENCTRL(id)) {}
+}
+
+static int _freqm_gpio_init(gpio_t msr_gpio_src, uint8_t *gclk_io_id)
+{
+    /* Check if selected pin has peripheral function GCLK */
+    int index = _freqm_pin(msr_gpio_src);
+
+    if (index < 0) {
+        return -1;
+    }
+
+    /* Initialize GPIO as input */
+    gpio_init(msr_gpio_src, GPIO_IN);
+    /* Enable peripheral function GCLK/IO on GPIO */
+    gpio_init_mux(msr_gpio_src, GPIO_MUX_M);
+    /* Lookup which GCLK_IO[x] must be used */
+    *gclk_io_id = gclk_io_ids[index];
+    /* Connect GCLK_IO[*gclk_io_id] with input pin */
+    _gclk_connect(*gclk_io_id, GCLK_SOURCE_GCLKIN, 0);
+
+    return 0;
+}
+
+static void _freqm_clock_init(uint8_t msr_clk_src, uint8_t ref_clk_src)
+{
+    /* Selection of the Generator and write Lock for FREQM_MSR */
+    GCLK->PCHCTRL[FREQM_GCLK_ID_MSR].reg = GCLK_PCHCTRL_GEN(msr_clk_src) | GCLK_PCHCTRL_CHEN_Msk;
+    /* Wait for synchronization */
+    while ((GCLK->PCHCTRL[FREQM_GCLK_ID_MSR].reg & GCLK_PCHCTRL_CHEN_Msk) != GCLK_PCHCTRL_CHEN_Msk) {}
+
+    /* Selection of the Generator and write Lock for FREQM_REF */
+    GCLK->PCHCTRL[FREQM_GCLK_ID_REF].reg = GCLK_PCHCTRL_GEN(ref_clk_src) | GCLK_PCHCTRL_CHEN_Msk;
+    /* Wait for synchronization */
+    while ((GCLK->PCHCTRL[FREQM_GCLK_ID_REF].reg & GCLK_PCHCTRL_CHEN_Msk) != GCLK_PCHCTRL_CHEN_Msk) {}
+}
+
+typedef struct {
+    freqm_callback_t callback;
+    void *context;
+} freqm_object_t;
+
+static volatile freqm_object_t freqm_obj;
+
+static mutex_t sync_lock = MUTEX_INIT_LOCKED;
+
+static void _freqm_enable(uint8_t refnum)
+{
+    FREQM->CFGA.reg = (uint16_t)(FREQM_CFGA_REFNUM(refnum));
+
+    /* Enable DONE Interrupt */
+    FREQM->INTENSET.reg = (uint8_t)FREQM_INTENSET_DONE_Msk;
+
+    /* Enable FREQM */
+    FREQM->CTRLA.reg = (uint8_t)FREQM_CTRLA_ENABLE_Msk;
+
+    /* Wait for Sync */
+    while ((FREQM->SYNCBUSY.reg) != 0U) {}
+}
+
+static bool _freqm_is_busy(void)
+{
+    bool freqm_is_busy = false;
+
+    if ((FREQM->STATUS.reg & FREQM_STATUS_BUSY_Msk) == FREQM_STATUS_BUSY_Msk) {
+        freqm_is_busy = true;
+    }
+
+    return freqm_is_busy;
+}
+
+static void _freqm_nvic_init(void)
+{
+    /* Enable interrupt */
+    NVIC_EnableIRQ(FREQM_IRQn);
+}
+
+bool freqm_start(int blocking)
+{
+    bool status = false;
+
+    /* Check if measurement is already in progress */
+    if (!_freqm_is_busy()) {
+        /* Clear the Done Interrupt flag */
+        FREQM->INTFLAG.reg = (uint8_t)FREQM_INTFLAG_DONE_Msk;
+
+        /* Start measurement */
+        FREQM->CTRLB.reg = (uint8_t)FREQM_CTRLB_START_Msk;
+
+        status = true;
+    }
+
+    if (blocking) {
+        /* Block until measurement is done */
+        mutex_lock(&sync_lock);
+    }
+
+    return status;
+}
+
+int freqm_frequency_get(const struct freqm_config *config, uint32_t *result)
```

Comment:
```suggestion
int freqm_frequency_get(freqm_t idx, uint32_t *result)
```

`struct freqm_config` is sam0 specific, better pass an index into the `freqm_config` array:

```C
const struct freqm_config *config = &freqm_config[idx];
```

with 

```C
/**
 * @brief   Define default Frequency meter type identifier
 */
#ifndef HAVE_FREQM_T
typedef uint_fast8_t freqm_t;
#endif
```

in `freqm.h`.